# Markov Property
## Definition
A state $S_t$ is *Markov* if and only if 
$$\mathbb{P}[S_{t+1}|S_t]=\mathbb{P}[S_{t+1}|S_t,\dots,S_t]$$
- The state captures all relevant information from the history
- Once the state is known, the history may be thrown away
- i.e. The state is a sufficient statistic of the future

# State Transition Matrix
For a Markov state $s$ and successor state $s'$, the *state transition probability* is defined by $$\mathcal{P}_{ss'}=\mathbb{P}[S_{t+1}=s'|S_t=s]$$ 
State transition matrix $\mathcal{P}$ defines transition probabilities from all states $s$ to all successor states $s'$,
$$\mathcal{P}=\mathrm{from} 
\begin{matrix}
    \mathrm{to}\\
    \begin{bmatrix}
    \mathcal{P}_{11},\dots,\mathcal{P}_{1n}\\
    \vdots,\ddots&\vdots\\
    \mathcal{P}_{n1}&\dots&\mathcal{P}_{nn}\\
    \end{bmatrix}
\end{matrix}$$
where each row of the matrix sums to 1.

# Markov Process
A Markov process is a memory less random process, random process, i.e. a sequence of random states $S_1,S_2,\cdots$ with the Markov property.
## Definition
A *Markov Process* (or *Markov Chain*) is a tuple $<\mathcal{S,P}>$
- $\mathcal{S}$ is a (finite) set of states
- $\mathcal{P}$ is a state transition probability matrix,
$$\mathcal{P}_{ss'}=\mathbb{P}[S_{t+1}=s'|S_t=s]$$

# Example: Student Markov Chain
![ExampleStudentMarkovChain](./pics/ExampleStudentMarkovChain.jpg)
Sample episodes for Student Markov Chain starting from $S_1=C1$
$$S_1,S_2,\cdots,S_T$$
- C1,C2,C3,Pass,Sleep
- C1,FB,FB,C1,C2,Sleep
- C1,C2,C3,Pub,C2,C3,Pass,Sleep
- C1,FB,FB,C1,C2,C3,Pub,C1,FB,FB,FB,C1,C2,C3,Pub,C2,Sleep
$$\begin{array}{lllllll}
{C1} & {C2} & {C3} & {Pass} & {Pub} & {FB} & {Sleep}
\end{array}$$
$$\mathcal{P}=\begin{bmatrix}
0 & 0.5 & 0 & 0 & 0 & 0.5 & 0\\
0 & 0 & 0.8 & 0 & 0 & 0 & 0.2\\
0 & 0 & 0 & 0.6 & 0.4 & 0 & 0\\
0 & 0 & 0 & 0 & 0 & 0 & 1\\
0.2 & 0.4 & 0.4 & 0 & 0 & 0 & 0 \\
0.1 & 0 & 0 & 0 & 0 & 0.9 & 0 \\
0 & 0 & 0 & 0 & 0 & 0 & 1\\
\end{bmatrix}$$

In [8]:
# Markov Property
# State Transition Matrix
# Example: Student Markov Chain
import numpy as np
# {C1} & {C2} & {C3} & {Pass} & {Pub} & {FB} & {Sleep}
P=np.matrix([[0 , 0.5 , 0 , 0 , 0 , 0.5 , 0],
[0 , 0 , 0.8 , 0 , 0 , 0 , 0.2],
[0 , 0 , 0 , 0.6 , 0.4 , 0 , 0],
[0 , 0 , 0 , 0 , 0 , 0 , 1],
[0.2 , 0.4 , 0.4 , 0 , 0 , 0 , 0],
[0.1 , 0 , 0 , 0 , 0 , 0.9 , 0],
[0 , 0 , 0 , 0 , 0 , 0 , 1]])
print("The state transition matrix (P) of Student Markov Chain is:")
print(P)

The state transition matrix (P) of Student Markov Chain is:
[[0.  0.5 0.  0.  0.  0.5 0. ]
 [0.  0.  0.8 0.  0.  0.  0.2]
 [0.  0.  0.  0.6 0.4 0.  0. ]
 [0.  0.  0.  0.  0.  0.  1. ]
 [0.2 0.4 0.4 0.  0.  0.  0. ]
 [0.1 0.  0.  0.  0.  0.9 0. ]
 [0.  0.  0.  0.  0.  0.  1. ]]


# Markov Reward Process
A Markov reward process is a Markov chain with values.
## Definition
A *Markov Reward Process* is a tuple $<\mathcal{S,P,{\color{red}R,\gamma}
}>$
- $\mathcal{S}$ is a finite set of states
- $\mathcal{P}$ is a state transition probability matrix,
- ${\color{red}\mathcal{R}}$ is a reward function, ${\color{red}\mathcal{R_s}=\mathbb{E}[R_{t+1}|S_{t}=s]}$
- ${\color{red}\gamma}$ is a discount factor, ${\color{red}\gamma\in[0,1]}$
## Example:
![ExampleStudentMarkovChain-MDP](./pics/ExampleStudentMarkovChain-MRP.jpg)
## Return
### Definition
The *return* $G_t$ is the total discounted reward from time-step $t$.
$$G_t=R_{t+1}+\gamma R_{t+2}+\dots=\sum_{k=0}^{\infty}{\gamma^kR_{t+k+1}}$$
- The *discount* $\gamma\in[0,1]$ is the present value of future rewards
- The value of receiving reward $R$ after $k+1$ time-steps is $\gamma^kR$
- This value immediate reward above delayed reward.
  - $\gamma$ close to 0 leads to "myopic" evaluation
  - $\gamma$ close to 1 leads to "far-sighted" evaluation

In [2]:
# Markov Reward Process
# A Markov reward process is a Markov chain with values.
## Return
### Definition
# The *return* $G_t$ is the total discounted reward from time-step $t$.
### Example: Student MRP Returns
gamma=0.5
# C1: C2: C3: Pass: Pub: FB: Sleep
R_s=np.array([-2,-2,-2,10,1,-1,0])
# C1,C2,C3,Pass,Sleep:
G1= R_s[0]+gamma*R_s[1]+gamma**2*R_s[2]+gamma**3*R_s[3]

# Solve it by define a function
R_dict={'C1':-2, 'C2':-2, 'C3':-2, 'Pass':10,'Pub':1,'FB':-1,'Sleep':0} # Reward dict
Route=[['C1','C2','C3','Pass','Sleep'], # Route list 1
['C1','FB','FB','C1','C2','Sleep'], # Route list 2
['C1','C2','C3','Pub','C2','C3','Pass','Sleep'], # Route list 3
['C1','FB','FB','C1','C2','C3','Pub','C1','FB','FB','FB','C1','C2','C3','Pub','C2','Sleep']] # Route list 4

def MRPReturn(Route,R_dict):
    "This fucntion calculate the Return based on Route and Reward dictionary"
    Return = 0
    for i in range(len(Route)):
        Return = Return + gamma**(i)*R_dict[Route[i]]
    return Return

Return=np.zeros(len(Route))
for i in range(len(Route)):
    Return[i]=MRPReturn(Route[i],R_dict)

# Print them
for i in range(len(Route)):
    print("Return of ",Route[i],":",Return[i])

Return of  ['C1', 'C2', 'C3', 'Pass', 'Sleep'] : -2.25
Return of  ['C1', 'FB', 'FB', 'C1', 'C2', 'Sleep'] : -3.125
Return of  ['C1', 'C2', 'C3', 'Pub', 'C2', 'C3', 'Pass', 'Sleep'] : -3.40625
Return of  ['C1', 'FB', 'FB', 'C1', 'C2', 'C3', 'Pub', 'C1', 'FB', 'FB', 'FB', 'C1', 'C2', 'C3', 'Pub', 'C2', 'Sleep'] : -3.196044921875


## Value Function
The value function $v(s)$ gives the long-term value of state $s$
### Definition
The *state valuefunction* $v(s)$ of an MRP is the expected *return* ($G_t$) starting from state $s$
$$v(s)=\mathbb{E}[G_t|S_t=s]$$
### Example: Student MRP Returns
Sample <font color=Red>returns</font> for Student MRP:
Starting from $S_1=C1$ with $\gamma=\frac{1}{2}$
$$G_1=R_2+\gamma R_3+\dots+\gamma^{T-2}R_T$$

- C1,C2,C3,Pass,Sleep:$$G_1=-2+\frac{1}{2}\times(-2)+\frac{1}{4}\times(-2)+\frac{1}{8}\times10=-2.25$$
- C1,FB,FB,C1,C2,Sleep:$$G_1=-2+\frac{1}{2}\times(-1)+\frac{1}{4}\times(-1)+\frac{1}{8}\times(-2)+\frac{1}{16}\times(-2)=-3.125$$
- C1,C2,C3,Pub,C2,C3,Pass,Sleep:...
- C1,FB,FB,C1,C2,C3,Pub,C1,FB,FB,FB,C1,C2,C3,Pub,C2,Sleep:...

### Example: State-Value Function for Student MRP(1)
![ExampleStudentMarkovChain-MDP-SVF](./pics/ExampleStudentMarkovChain-MRP-State-value-function1.jpg)
### Example: State-Value Function for Student MRP(1)
![ExampleStudentMarkovChain-MDP-SVF](./pics/ExampleStudentMarkovChain-MRP-State-value-function2.jpg)
### Example: State-Value Function for Student MRP(1)
![ExampleStudentMarkovChain-MDP-SVF](./pics/ExampleStudentMarkovChain-MRP-State-value-function3.jpg)

## Bellman Equation for MRPs
The value function can be decomposed into two parts:
- immediate reward $R_{t+1}$
- discounted value of successor state $\gamma v(S_{t+1})$
$$
\begin{align}
v(s) & =\mathbb{E}[G_t|S_t=s]\\
& =\mathbb{E}[R_{t+1}+\gamma R_{t+2}+\gamma^2R_{t+3}+\dots|S_t=s]\\
& =\mathbb{E}[R_{t+1}+\gamma(R_{t+2}+\gamma R_{t+3}+\dots)|S_t=s]\\
& =\mathbb{E}[R_{t+1}+\gamma G_{t+1}|S_t=s]\\
& =\mathbb{E}[R_{t+1}+\gamma v(S_{t+1})|S_t=s]
\end{align}
$$
$$
v(s)=\mathcal{R_s}+\gamma \sum_{s'\in\mathcal{S}}{\mathcal{P_{ss'}}v(s')}
$$

### Example:Bellman Equation for Student MRP
![Bellman](./pics/BellmanEquationExample.jpg)
### Bellman Equation in Matrix Form
The Bellman equation can be expressed concisely using matrices,
$$\mathbf{v}=\mathbf{\mathcal{R}}+\gamma \mathcal{P}\mathbf{v}$$
where $\mathbf{v}$ is a column vector with one entry per state
$$\begin{bmatrix}v(1)\\\vdots\\v(n)\end{bmatrix}=\begin{bmatrix}\mathcal{R}_1\\\vdots\\\mathcal{R}_n\end{bmatrix}+\gamma \begin{bmatrix}\mathcal{P}_{11}&\dots&\mathcal{P}_{1n}\\\vdots\\\mathcal{P}_{n1}&\dots&\mathcal{P}_{nn}\end{bmatrix}\begin{bmatrix}v(1)\\\vdots\\v(n)\end{bmatrix}$$
### Solving the Bellman Equation
- The Bellman equation is a linear equation 
- It can be solved directly:
$$
\mathbf{v}=\mathbf{\mathcal{R}}+\gamma \mathcal{P}\mathbf{v}→
\mathbf{v}=(I-\gamma \mathcal{P})^{-1}\mathbf{\mathcal{R}}
$$
- Computational complexity is $O(n^3)$ for $n$ states
- Direct solution only possible for small MRPs
- There are many iterative methods for large MRPs, e.g.
  - Dynamic programming
  - Monte-Carlo evaluation
  - Temporal-Difference learning

In [14]:
## Value Function
# The value function $v(s)$ gives the long-term value of state $s$
### Definition
# The *state valuefunction* $v(s)$ of an MRP is the expected *return* ($G_t$) starting from state $s$

# Solve it by define a function
Value_dict={'C1':-13, 'C2':1.5, 'C3':4.3, 'Pass':10,'Pub':0.8,'FB':-23,'Sleep':0} # Value function dict
print("Value of State C3: ",Value_dict['C3'])
print("Value of State C3 calculated by Bellman Equation (Gamma=1): ", R_dict['C3']+P[2,3]*Value_dict['Pass']+P[2,4]*Value_dict['Pub'])

## Bellman Equation for MRPs
### Example:Bellman Equation for Student MRP
l=len(Value_dict)

R_array = np.array(list(R_dict.values()))
R_array = np.array([R_array])

# print(np.linalg.inv(np.identity(l)-gamma*P))
gamma2 = 0.999999 # when gamma2=1, the inverse matrix is singular, WHY???
V=np.linalg.inv(np.identity(l)-gamma2*P)*(R_array.T)
print("When Gamma is ",gamma2," :")
print(V)


Value of State C3:  4.3
Value of State C3 calculated by Bellman Equation (Gamma=1):  4.32
When Gamma is  0.999999  :
[[-12.54296219]
 [  1.4568013 ]
 [  4.32100594]
 [ 10.        ]
 [  0.80253065]
 [-22.54274676]
 [  0.        ]]


# Markov Decision Process
A Markov decision process (MDP) is a Markov reward process with decisions. It is an *environment* in which all states are Markov.
## Definition
A *Markov Decision Process* is tuple $<\mathcal{S},{\color{red}\mathcal{A}},\mathcal{P},\mathcal{R},\mathcal{\gamma}>$
- $\mathcal{S}$ is a finite set of states
- <font color=Red> $\mathcal{A}$ is a finite set of actitions</font>
- $\mathcal{P}$ is a state transition probability matrix,
  $$\mathcal{P}_{ss'}^{\color{red}a}=\mathbb{P}[S_{t+1}=s'|S_t=s,A_t={\color{red}a}]$$
- $\mathcal{R}$ is a reward function, $\mathcal{R}_s^{\color{red}a}=\mathbb{E}[R_{t+1}|S_t=s,A_t={\color{red}a}]$
- $\gamma$ is a discount factor $\gamma\in[0,1]$

## Example: Student MDP
![ExampleStudentMarkovDecisionProcess](./pics/ExampleStudentMarkovDecisionProcess.jpg)

## Policies
### Definition
A *policy* $\pi$ is a distribution over actions given states,
$$\pi(a|s)=\mathbb{P}[A_t=a|S_t=s]$$
- A policy fully defines the behaviour of an agent
- MDP policies depend on the current state (not the history)
- i.e. Policies are *stationary* (time-independent),
$$A_t\sim\pi(·|S_t),\forall t>0 $$
- Given an MDP $M=\langle \mathcal{S,A,P,R,\gamma} \rangle$ and a policy $\pi$
- The state sequnce $S_1,S_2,\dots$ is a Markov process $\langle \mathcal{S,P^{\pi}}\rangle$
- The state and reward sequnce $S_1,R_2,S_2,\dots$ is a Markov reward process $\langle \mathcal{S,P^{\pi},R^{\pi},\gamma} \rangle$
- where
$$\mathcal{P}^{\pi}_{s,s'}=\sum_{a\in\mathcal{A}}{\pi(a|s)\mathcal{P}^{a}_{s,s'}}\\\mathcal{R}^{\pi}_{s}=\sum_{a\in\mathcal{A}}{\pi(a|s)\mathcal{R}^{a}_{s}}$$

## Value Function
### Definition
The ***state-value function*** $v_\pi(s)$ of an MDP is the expected return starting from state $s$, and then following policy $\pi$
$$v_\pi(s)=\mathbb{E}_\pi[G_t|S_t=s]$$
### Definition
The ***action-value function*** $q_\pi(s,a)$ of an MDP is the expected return starting from state $s$, taking action $a$ and then following policy $\pi$
$$q_\pi(s,a)=\mathbb{E}_\pi[G_t|S_t=s,A_t=a]$$
### Example: State-Value Function for Student MDP
![ExampleStudentMarkovDecisionProcess_SVF](./pics/ExampleStudentMarkovDecisionProcess_StateValueFunction.jpg)

## Bellman Expectation Equation
The state-value function can again be decomposed into immediate reward plus discounted value of successor state,
$$v_\pi(s)=\mathbb{E}_\pi[R_{t+1}+\gamma v_\pi(S_{t+1}|S_t=s)]$$
The action-value function can similarly be decomposed,
$$q_\pi(s,a)=\mathbb{E}_\pi[R_{t+1}+\gamma q_\pi(S_{t+1},A_{t+1}|S_t=s,A_t=a)]$$
### Bellman Expectation Equation for $V_\pi,Q_\pi$
$$
v_\pi(s)=\sum_{a\in\mathcal{A}}{\pi(a|s)q_\pi(s,a)}\\
q_\pi(s,a)=\mathcal{R}_s^a+\gamma \sum_{s'\in\mathcal{S}}{\mathcal{P}^{a}_{ss'}v_\pi(s')}\\
v_\pi(s)=\sum_{a\in\mathcal{A}}{\pi(a|s)(\mathcal{R}_s^a+\gamma \sum_{s'\in\mathcal{S}}{\mathcal{P}^{a}_{ss'}v_\pi(s')})}\\
q_\pi(s,a)=\mathcal{R}_s^a+\gamma \sum_{s'\in\mathcal{S}}{\mathcal{P}^{a}_{ss'}\sum_{a'\in\mathcal{A}}{\pi(a'|s')q_\pi(s',a')}}\\
$$
### Example: Bellman Expectation Equation in Student MDP
![ExampleStudentMarkovDecisionProcessSVF_BellmanExpEq](./pics/ExampleStudentMarkovDecisionProcess_StateValueFunction_BellmanExpectationEq.jpg)
### Bellman Expectation Equation (Matrix Form)
The Bellman expectation equation can be expressed concisely using the induced MRP,
$$v_{\pi} =\mathcal{R^\pi+\gamma P^\pi v_\pi}$$
with direct solution
$$v_{\pi} =(I-\mathcal{\gamma P^\pi})^{-1}\mathcal{R^\pi}$$

In [23]:
# Markov Decision Process
## Policies
## Value Function: state-value function and action-value function
## Example: Student MDP
# v_π(s) for π(a|s)=0.5, γ =1
π_as=0.5
gamma3=1
Vpi={'C1':-1.3, 'C2':2.7, 'C3':7.4, 'Sleep':0,'FB':-2.3} # State-Value function dict
# 'C1', 'C2', 'C3', 'Sleep','FB'
 # C1→Study→C2: 0.5, C1→FB(action)→FB(state): 0.5
Ppi_MDP=np.matrix([[0 , π_as , 0 , 0 , 1-π_as], \
 # C2→Study(action)→C3: 0.5, C2→Sleep(action)→Sleep(state): 0.5
[0 , 0 , π_as , 1-π_as , 0], \
# C3→Study(action)→Sleep(state): 0.5, C3→Pub(action)→C1: 0.5*0.2, C3→Pub(action)→C2/C3: 0.5*0.4
[π_as*0.2 , π_as*0.4 , π_as*0.4 , π_as, 0], \
 # Sleep is the end state
[0 , 0 , 0 , 0 , 0], \
# FB(state)→FB(action)→FB(state): 0.5, FB(state)→Quit→FB(state): 0.5
[π_as , 0 , 0, 0, 1-π_as]]) 
R_sa={'C1Study':-2,'C1FB':-1,'FBFB':-1,'FBC1':0,'C2Study':-2,'C2Sleep':0,'C3Study':10,'C3Pub':1} #Reward dict
R_spi=np.array([π_as*R_sa['C1Study']+(1-π_as)*R_sa['C1FB'], \
    π_as*R_sa['C2Study']+(1-π_as)*R_sa['C2Sleep'], \
    π_as*R_sa['C3Study']+(1-π_as)*R_sa['C3Pub'], \
    0, \
    π_as*R_sa['FBC1']+(1-π_as)*R_sa['FBFB']])

### Example: Bellman Expectation Equation in Student MDP
print(Vpi['C3'])
print('State Value of "C1" according to Bellman Expectation Equation:')
v_C3=π_as*(R_sa['C3Study']+R_sa['C3Pub']+gamma3*(0.2*Vpi['C1']+0.4*Vpi['C2']+0.4*Vpi['C3']))
print(v_C3)

### Example: Bellman Expectation Equation (Matrix Form) (not confirmed yet)
gamma3 = 1
R_spi=np.array([R_spi])
l=len(R_spi)
Vpi_value=np.linalg.inv(np.identity(l)-gamma3*Ppi_MDP)*(R_spi.T)
print("When Gamma is ",gamma3," :")
print(Vpi_value)



7.4
State Value of "C1" according to Bellman Expectation Equation:
7.390000000000001
When Gamma is  1  :
[[ -5.]
 [ -3.]
 [ 18.]
 [-16.]
 [  6.]]


# Optimal Value Fuction
## Definition
The *optimal state-value funtion* $v_*(s)$ is the maximum value function over all policies
$$v_*(s)=\max_{\pi} v_\pi(s)$$
The *optimal action-value function* $q_*(s,a)$ is the maximum action-value function over all policies
$$q_*(s,a)=\max_{\pi} q_*(s,a)$$
- The optimal value function specifies the best possible performance in the MDP.
- An MDP is "solved" when we know the optimal value fn. 